### ISSUE TITLE
Add Security module for Azure AD

### ISSUE THREAD

**Issue by tidyui** [PiranhaCMS/piranha.core#267]
Create security module for Azure AD so that the CMS can be easily integrated into an existing user base

**Comment by lestersconyers** [comment]
Is this enhancement for Azure AD + Piranha Identity?

**Comment by tidyui** [comment]
I haven't had the time to investigate it yet, but the idea is to.

1. Implement the `ISecurity` to authenticate users against Azure AD
2. If possible, implement admin views to view available users

This isn't really my area of expertise so I need to do some investigation on how it's done. Also the users from the AD needs to be connected with the claims they need for the Piranha Manager, not really sure if this kind of thing can be administered from the Azure AD or if you actually need a local Identity user handling these things and you only use the AD for login.

Do you have any experience in integrating Azure AD with .NET Core?

**Comment by lestersconyers** [comment]
Yes, I've integrated Piranha with Azure AD as well. Since you have the manager set up with claims it's relatively straightforward if you augment the claims within startup code. Mapping custom claims in Azure AD is a pain so I would avoid it. Ideally those claims would be augmented by the roles assigned in the manager. I haven't gotten that far yet but would love to help.

**Comment by tidyui** [comment]
Well I haven't got that much experience with any of this :) How does it work, are you using the Identity module but with a custom `ISecurity` implementation that authenticates with Azure AD. Then if the user is authenticated you map it to a local Identity user, and if it doesn't exist you create it? Are you using the standard login view for the manager or are you redirecting to a single-sing-on view from Azure AD?

Regards

**Comment by lestersconyers** [comment]
I followed these steps to setup WSFederation using Azure AD. https://docs.microsoft.com/en-us/aspnet/core/security/authentication/ws-federation?view=aspnetcore-2.1#use-ws-federation-without-aspnet-core-identity. So the basics it really had nothing to do with Piranha. I'll fork your examples repo and submit a PR for you to review as a basic example.

**Comment by tidyui** [comment]
Hi again @lestersconyers Do you have some code regarding this that you might be willing to share? Got a question regarding this from another user 🙂

**Comment by filipelteixeira** [comment]
Hey I’m evaluating using piranha CMS as our new code base for a public website, but it is kind of a requirement to use azure ad for the manager area, if you could share some example code I would pick it from there.
Thanks

**Comment by eloekset** [comment]
I need support for IdentityServer. Both Azure AD and IdentityServer are OAuth2, so there shouldn't be much of a difference. I'll try it out in my fork and share if I get it up and running. (Also need multi-tenancy BTW, as mentioned in #1132 and related, so I'll see how far I get using Piranha.)

**Comment by tidyui** [comment]
@eloekset We really don’t have full multi tenancy planned as this is currently not in the roadmap.

Best regards

**Comment by eloekset** [comment]
Yes, I saw that. I will have to live in a fork and keep it updated from your master branch. My first goal is OpenID/OAuth2 with IdSrv/Azure AD, and then it shouldn't be too hard to extend all storage with an AccountID or TenantID.

**Comment by eloekset** [comment]
I've finally had some time to work on the Azure AD module again, but I'm struggling with an issue that is difficult to figure out, so I created a discussion over in the  aspnetcore repo and hope to get some help by someone: https://github.com/dotnet/aspnetcore/discussions/24690

**Comment by vigouredelaruse** [comment]
i've also been trying to authenticate with azure ad b2c and i have this working. 

by working i mean
- anonymous get /manager redirects to /microsoftidentity/account/signin 
- azure ad b2c challenge is displayed, with identity provided by social. logins and azure ad
- people who aren't in a specifically configured set of azure ad tenants cannot authenticate to /manager - this means the social logins don't get to authenticate at /manager even though if they type /manager they will get an auth challenge
-  on. login from an approved azure ad tenant the manager interface is displayed
- identitywithseed is being used, to hack around dependency injection issues if there is no identity service registered. this means that in certain circumstances you do see the identitywithseed challange despite my efforts to date. but it's only visible if you're logged in to an approved azure ad tenant. the problem with the identity interface as i see it is the username/password on the interface definition. the impetus. for doing this is to keep passwords out of the database

how i did it
- if anyone responds here i will tell you. other than that i will publish a blog and update this with the link later. i expect you know i will see you hitting /manager and in the azure risky behaviour analytics
- essentially i forwarded challenges, loginurls for cookie auth to the oidc scheme registered by microsoftidentity
- also i wrote an asp.net core iclaimstransformer that supplies the full set of admin claims to approved users. therefore this solution is incomplete, in that there is no mapping mechanism between azure rbac (https://docs.microsoft.com/en-us/azure/role-based-access-control/overview) (b2c groups are a special case https://mrochon.azurewebsites.net/2019/05/06/using-groups-in-azure-ad-b2c/) and piranha rbac- for that matter, i hardcoded the approved. domains, so there. obviously. a production attempt would read such things from asp.net core configuration

**Comment by tedvanderveen** [comment]
@vigouredelaruse maybe #1623 can get you going more efficiently?
